posts/2024-09-05-impl-trait-capture-rules.md

```diff
@@ -0,0 +1,211 @@
+
+## Rust 2024 design
+
+The above problems motivated us to take a new approach in Rust 2024. The approach is a combination of two things:
+
+* a new default, applicable only in Rust 2024;
+* a syntax for explicit capture, usably in any edition, written like 
+
+We cover each in turn.
+
+### Impl Traits now capture lifetimes by default
+
+In Rust 2024, the default is that return-position impl Trait values can capture all the references that are in scope. Therefore the initial example of this blog post will compile just fine in Rust 2024 ([try it yourself](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=d366396da2fbd5334b7560c3dfb3290b)):
+
+```rust
+fn process_datums(
+    datums: &[Datum]
+) -> impl Iterator<Item = ProcessedDatum> {
+    datums
+        .iter()
+        .map(|datum| datum.process())
+}
+```
+
+Yay!
+
+### Impl Traits can include a `use<>` bound to specify precisely which generic types and lifetimes they use
+
+There are however times where it is useful not to capture lifetime parameters. Consider this example, which takes a slice of type T `&[T]` and returns an iterator over its indices:
+
+```rust!
+fn indices<T>(
+    slice: &[T],
+) -> impl Iterator<Item = usize> {
+    0 .. slice.len()
+}
+```
+
+If you look closely, you can see that it only needs to read the length of the slice and doesn't hold a reference to `slice` afterwards. Nonetheless, callers today will assume that the return value *may* hold a reference to `slice`, which means that this caller for example would get an error:
+
+```rust
+fn main() {
+    let mut data = vec![1, 2, 3];
+    let i = indices(&data);
+    data.push(4); // <-- error!
+    i.next();
+}
+```
+
+This may actually be what you want: it gives you room to modify `indices` later and have it actually make use of `slice` in the returned iterator, for example. Put another way, the new default continues the `impl Trait` tradition of retaining flexibility for the function to change its implementation without breaking callers.
+
+But what if it's *not* what you want? What if you want to guarantee that `indices` will not retain a reference to `data` in its return value? You can do that now with the new `use` bounds. The idea is that your `impl Trait` can include a bound written `use<...>` which lists out the generic parameters that the return value may use. If you include `+ use<>`, then, this means "do not capture anything". So changing `indices` to be written as follows will allow `main` to compile:
+
+```rust
+fn indices<T>(
+    slice: &[T],
+) -> impl Iterator<Item = usize> + use<> {
```

Comment:
I'm sure this is far too late for input, but I'm a bit let down that despite using the word `capture` approximately 30 times, including in the PR title, URL, headers, and all over the text...the keyword chosen was `use`.  I'm not looking forward to having to teach my students to remember this special way to use `use` to specify what to _capture_. 🤦🏻‍♂️ 😜

Reply:
Appreciate the feedback! Actually, maybe I should try to avoid saying capture, because the intuition really isn't as good as "using". That is, the word "use" was chosen because it corresponds to what values are *used* by the hidden type (that is, appear in). Capture seems to imply that things are 'taken' from the place they appear.

Reply:
Using more consistent language between the code and the description would definitely help with comprehension.

Reply:
I made the edits,  @CleanCut, take a look at the new revision. I felt it overall reads much better.